### ISSUE TITLE
Does starlette support arrays as form/queryparam input?

### ISSUE THREAD

**Issue by jodusan** [Kludex/starlette#678]
Is it possible to send array within form or query param?

```
<!DOCTYPE html>
<html>
<body>

<h2>HTML Forms</h2>

<!-- Simple form which will send a POST request -->
<form action="http://localhost:8000/test" method="post">
  <label for="POST-name">Name:</label>
<input type="text" name="comment[]" value="comment1"/>
<input type="text" name="comment[]" value="comment2"/>
<input type="text" name="comment[]" value="comment3"/>
<input type="text" name="comment[]" value="comment4"/>

  <input type="submit" value="Save">
</form>
</body>
</html>
```

Posting this form results in a string rather than array:
![image](https://user-images.githubusercontent.com/2333381/66778279-80e87300-eecb-11e9-84e3-75867a6043d6.png)

**Comment by lovelydinosaur** [comment]
If it doesn't then it ought to.

Our tests for body data are here:

https://github.com/encode/starlette/blob/master/tests/test_formparsers.py

There's two ways it can be encoded - urlencoded or multipart. It looks like we've got a test cases for multiple items under the same key with multipart, but not a case case for urlencoded.

**Comment by lovelydinosaur** [comment]
It *looks* to me like the parser handles it correctly... https://github.com/encode/starlette/blob/master/starlette/formparsers.py

You probably want to check by using a command line tool to upload some data, or have a look in the network inspector to make sure you're *actually* sending multiple key/value pairs.

**Comment by lovelydinosaur** [comment]
Closing this on the assumption that there's not an issue here, but happy to reopen if needed.

**Comment by cass-e-design** [comment]
I'm adding to this since I found this thread while encountering this exact issue:

in 0.13.4 (and presumably earlier) when you process multi-part / x-www-form-urlencoded with `request.form()` it returns [`FormData`](https://github.com/encode/starlette/blob/master/starlette/datastructures.py). 
It's true that in `_dict` or `FormData['arrayname[]']` it only keeps the last array member. In the original example, it only keeps `{'comment[]': 'comment4'}`.
However, it keeps every element in its `_list` of tuples, ie `[('comment[]', 'comment1'), ('comment[]', 'comment2')]` etc, and if you call `FormData.getlist()` it returns the full list for you.

TLDR: Call `FormData.getlist('arrayname[]')` for the full list.

**Comment by invokermain** [comment]
FYI: not knowledgeable about this area just creating discussion. And this is all for the get request/query parameter case. Maybe I'm misinformed...

Is this handled correctly? The lack of documentation around this makes it feel like it's not handled. With a query string like `somevar[]=1&somevar[]=2`, what should the user expect?

It seems surprising to me that the answer is: `request.query_params["somevar[]"]  #... 2`. I mean, where did the 1 go? I imagine in most cases what the developer wants is `request.query_params["somevar"]  #... (1, 2)`. 

@cass-e-design's answer seems non-obvious. If that is the recommended way of handling this type of query_params should it not be documented, seems complicated enough to warrant it?

**EDIT**
Doing some further reading into this it looks like the "[]" syntax (or any way of passing arrays via query parameters in general) is not an accepted standard. So can see why it's not treated specially.

Still would be nice to see Starlette have an opinion on this? Seems like an unintended consequence of using ImmutableDict for QueryParameters that duplicate parameters are ignored: `self._dict = {k: v for k, v in _items}`. Could Starlette support some sort of  way of passing an array parameter or should we always just pass as a comma separated list and parse into a list manually ourselves?

**Working Solution**
```
    def _handle_array_query_params(params: QueryParams) -> QueryParams:
        array_params = (param for param in params._dict.copy() if param.endswith("[]"))
        for param in array_params:
            params._dict[param[:-2]] = params.getlist(param)
            del params._dict[param]
        return params
```

**Comment by jdhao** [comment]
Seems like we can use `request.query_params.multi_items()` to get the key-val tuple as a list. Then we can parse it whatever we want.

Another way is to use `query_params.getlist(<some-key>)` to retrieve the value, but then all values become a list, even for query param that only has one value. But I guess it is not a big issue?

**Comment by ChannyClaus** [comment]
```
>> from urllib.parse import parse_qs, urlparse
>> parse_qs(urlparse(str(request.url)).query)
{'note_id': ['1234', '1244'], 'region_id': ['1234']}
```
if the desired result is `key -> array of values`, this seems to work as well.
